### ISSUE TITLE
io::driver should have loom tests

### ISSUE THREAD

**Issue by bdonlan** [tokio-rs/tokio#3018]
There are some bugs such as #2924 that would be much easier to test with loom testing. This will however require some internal refactoring and mocking to isolate the IO driver from mio and the underlying OS.

**Comment by darkraider01** [comment]
Hi! I’m interested in working on this and took a look at the current implementation. runtime::io::Driver still directly owns the Mio poller and event buffers, while Handle owns the Mio registry and waker. The registration lifecycle also relies on deferred release so ScheduledIo remains alive while polling.
Since this issue is several years old, I wanted to check whether making this area testable with Loom is still something the project wants. If so, would a small internal abstraction around polling and event delivery still be a reasonable first step, or is there a different testing direction you’d prefer today?
I’d be happy to put together a more concrete design before starting a PR, would appreciate any guidance before I start, since this touches fairly sensitive internals!!

**Comment by Darksonn** [comment]
If we can do this in a way that doesn't require too much refactoring, and the loom tests don't take too long to run, then I'm open to it.

**Comment by darkraider01** [comment]
I’ll start with a small prototype focused on the deregistration/event-delivery race, keep the refactoring minimal, and measure the Loom test runtime. I’ll share the approach before expanding the scope.
